File: app/taskMgr/src/main/java/org/exthmui/share/taskMgr/entities/TaskEntity.java
package org.exthmui.share.taskMgr.entities;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import org.exthmui.share.taskMgr.Task;
import org.exthmui.share.taskMgr.TaskStatus;

import java.util.concurrent.ExecutionException;

@Entity(tableName = "task")
public class TaskEntity {
    @PrimaryKey
    @NonNull
    @ColumnInfo(name = "task_id")
    private final String taskId;

    @ColumnInfo(name = "group_id")
    private final String groupId;

    @ColumnInfo(name = "task_type")
    private final String taskType;

    @ColumnInfo(name = "status")
    private TaskStatus status;

    @ColumnInfo(name = "input_data")
    private final Bundle inputData;

    @ColumnInfo(name = "result_id")
    private String resultId;

    @ColumnInfo(name = "progress_data")
    private volatile Bundle progressData;

    public TaskEntity(@NonNull String taskId, String groupId, String taskType, TaskStatus status, Bundle inputData, String resultId, Bundle progressData) {
        this.taskId = taskId;
        this.groupId = groupId;
        this.taskType = taskType;
        this.status = status;
        this.inputData = inputData;
        this.resultId = resultId;
        this.progressData = progressData;
    }

    public TaskEntity(@NonNull Task task, String groupId) {
        this.taskId = task.getTaskId();
        this.groupId = groupId;
        this.taskType = task.getClass().getCanonicalName();
        this.inputData = task.getInputData();
        updateTask(task);
    }

    public void updateTask(Task task) {
        this.status = task.getStatus();
        try {
            this.resultId = task.getResultFuture().isDone() ? task.getResultFuture().get().getResultId() : null;
        } catch (ExecutionException | InterruptedException e) {
            throw new RuntimeException(e);
        }
        this.progressData = task.getProgressData();
    }
    @NonNull
    public String getTaskId() {
        return taskId;
    }

    public String getGroupId() {
        return groupId;
    }

    public String getTaskType() {
        return taskType;
    }

    public TaskStatus getStatus() {
        return status;
    }

    public void setStatus(TaskStatus status) {
        this.status = status;
    }

    public Bundle getInputData() {
        return inputData;
    }

    public String getResultId() {
        return resultId;
    }

    public void setResult(String resultId) {
        this.resultId = resultId;
    }

    public Bundle getProgressData() {
        return progressData;
    }

    public void setProgressData(Bundle progressData) {
        this.progressData = progressData;
    }
}

File: app/taskMgr/src/main/java/org/exthmui/share/taskMgr/entities/TaskEntity.kt
package org.exthmui.share.taskMgr.entities

import android.os.Bundle
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import org.exthmui.share.taskMgr.Task
import org.exthmui.share.taskMgr.TaskStatus
import org.exthmui.share.taskMgr.converters.BundleConverter
import java.util.concurrent.ExecutionException

@Entity(tableName = "task")
class TaskEntity {
    @PrimaryKey
    @ColumnInfo(name = "task_id")
    val taskId: String

    @ColumnInfo(name = "group_id")
    val groupId: String

    @ColumnInfo(name = "task_type")
    val taskType: String

    @ColumnInfo(name = "status")
    var status: TaskStatus = TaskStatus.CREATED

    @ColumnInfo(name = "input_data")
    @TypeConverters(BundleConverter::class)
    val inputData: Bundle

    @ColumnInfo(name = "result_id")
    var resultId: String? = null
        private set

    @ColumnInfo(name = "progress_data")
    @TypeConverters(BundleConverter::class)
    @Volatile
    var progressData: Bundle? = null

    constructor(
        taskId: String,
        groupId: String,
        taskType: String,
        status: TaskStatus,
        inputData: Bundle,
        resultId: String?,
        progressData: Bundle?
    ) {
        this.taskId = taskId
        this.groupId = groupId
        this.taskType = taskType
        this.status = status
        this.inputData = inputData
        this.resultId = resultId
        this.progressData = progressData
    }

    constructor(task: Task, groupId: String) {
        taskId = task.taskId
        this.groupId = groupId
        taskType = task.javaClass.canonicalName ?: ""
        inputData = task.inputData
        updateTask(task)
    }

    fun updateTask(task: Task) {
        status = task.status
        try {
            resultId = if (task.resultFuture.isDone) task.resultFuture.get()!!.resultId else null
        } catch (e: ExecutionException) {
            throw RuntimeException(e)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
        progressData = task.progressData
    }

    fun setResult(resultId: String?) {
        this.resultId = resultId
    }
}